core/src/main/java/com/linecorp/armeria/client/endpoint/RampingUpWeightedRoundRobinStrategy.java

```diff
@@ -0,0 +1,437 @@
+                if (canAddToPrevEntry()) {
+                    // Update weight right away and combine updating weight schedule with the previous Entry.
+                    final Set<EndpointAndStep> newlyAddedEndpoints = updateEndpoints1(newEndpoints);
+                    final ImmutableList.Builder<Endpoint> builder = ImmutableList.builder();
+                    addCurrentEndpoints(builder);
+
+                    if (!newlyAddedEndpoints.isEmpty()) {
+                        updateEndpointWeight(newlyAddedEndpoints, builder);
+                        endpointsInUpdatingEntries.getLast().addEndpoints(newlyAddedEndpoints);
+                    }
+                    // Should recreate endpointSelector even when newlyAddedEndpoints is empty because
+                    // the oldEndpoints is changed.
+                    endpointSelector = new WeightBasedRandomEndpointSelector(builder.build());
+                    return;
+                }
+
+                if (canAddToNextEntry()) {
+                    // Combine with the next entry and let it handle this newEndpoints when
+                    // updateEndpointWeight() is executed.
+                    unhandledNewEndpoints = newEndpoints;
+                    return;
+                }
+            }
+
+            final Set<EndpointAndStep> newlyAddedEndpoints = updateEndpoints1(newEndpoints);
+            if (newlyAddedEndpoints.isEmpty()) {
+                final ImmutableList.Builder<Endpoint> builder = ImmutableList.builder();
+                addCurrentEndpoints(builder);
+                // Should recreate endpointSelector because the oldEndpoints is changed
+                // when newlyAddedEndpoints is empty.
+                endpointSelector = new WeightBasedRandomEndpointSelector(builder.build());
+                return;
+            }
+
+            final EndpointsInUpdatingEntry endpointsInUpdatingEntry =
+                    new EndpointsInUpdatingEntry(newlyAddedEndpoints, ticker,
+                                                 TimeUnit.MILLISECONDS.toNanos(rampingUpIntervalMillis));
+            endpointsInUpdatingEntries.add(endpointsInUpdatingEntry);
+            updateEndpointWeight();
+        }
+
+        private boolean canAddToPrevEntry() {
+            final EndpointsInUpdatingEntry lastEndpointsInUpdatingEntry = endpointsInUpdatingEntries.peekLast();
+            return lastEndpointsInUpdatingEntry != null &&
+                   ticker.read() - lastEndpointsInUpdatingEntry.lastUpdatedTime <= updatingTaskWindowNanos;
+        }
+
+        private boolean canAddToNextEntry() {
+            final EndpointsInUpdatingEntry nextEndpointsInUpdatingEntry = endpointsInUpdatingEntries.peek();
+            return nextEndpointsInUpdatingEntry != null &&
+                   nextEndpointsInUpdatingEntry.nextUpdatingTime - ticker.read() <= updatingTaskWindowNanos;
+        }
+
+        private void addCurrentEndpoints(Builder<Endpoint> builder) {
```

Comment:
How about returning a mutable `List` here? The created immutable list of `Endpoint` is copied again in the constructor of `WeightBasedRandomEndpointSelector`.

Reply:
Changed the name to `targetEndpointsBuilder` as we discussed. Thanks!